sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -639,6 +640,27 @@ object HiveTypeCoercion {
   }
 
   /**
+   * Turns Add/Subtract of DateType/TimestampType/StringType and CalendarIntervalType
+   * to TimeAdd/TimeSub
+   */
+  object DateTimeOperations extends Rule[LogicalPlan] {
```

Comment:
can you add a unit test for this rule? 

make sure there is one test case just for this rule, and another that tests DateTimeOperations and ImplicitTypeCasts in combination to make sure they still work. I worry if we change ImplicitTypeCasts in the future, DateTimeOperations might break (e.g. the interval got converted to string type)